tools/api/generate_go_protobuf.py

```diff
@@ -0,0 +1,112 @@
+
+TARGETS = '@envoy_api//...'
+IMPORT_BASE = 'github.com/envoyproxy/go-control-plane'
+OUTPUT_BASE = 'build_go'
+REPO_BASE = 'go-control-plane'
+BRANCH = 'master'
+MIRROR_MSG = 'Mirrored from envoyproxy/envoy @ '
+USER_NAME = 'Kuat Yessenov'
+USER_EMAIL = 'kuat@google.com'
+
+
+def generateProtobufs(output):
+  bazel_bin = check_output(['bazel', 'info', 'bazel-bin']).decode().strip()
+  go_protos = check_output([
+      'bazel',
+      'query',
+      'kind("go_proto_library", %s)' % TARGETS,
+  ]).split()
+
+  # Each rule has the form @envoy_api//foo/bar:baz_go_proto.
+  # First build all the rules to ensure we have the output files.
+  check_call(['bazel', 'build', '-c', 'fastbuild'] + go_protos)
+
+  shutil.rmtree(output, ignore_errors=True)
+  for rule in go_protos:
+    # Example rule:
+    # @envoy_api//envoy/config/bootstrap/v2:pkg_go_proto
+    #
+    # Example generated directory:
+    # bazel-bin/external/envoy_api/envoy/config/bootstrap/v2/linux_amd64_stripped/pkg_go_proto%/github.com/envoyproxy/go-control-plane/envoy/config/bootstrap/v2/
+    #
+    # Example output directory:
+    # go_out/envoy/config/bootstrap/v2
+    rule_dir, proto = rule.decode()[len('@envoy_api//'):].rsplit(':', 1)
+
+    input_dir = os.path.join(bazel_bin, 'external', 'envoy_api', rule_dir, 'linux_amd64_stripped',
+                             proto + '%', IMPORT_BASE, rule_dir)
+    input_files = glob.glob(os.path.join(input_dir, '*.go'))
+    output_dir = os.path.join(output, rule_dir)
+
+    # Ensure the output directory exists
+    os.makedirs(output_dir, 0o755, exist_ok=True)
+    for generated_file in input_files:
+      shutil.copy(generated_file, output_dir)
+  print('Go artifacts placed into: ' + output)
+
+
+def cloneGoProtobufs(repo):
+  # Create a local clone of go-control-plane
+  shutil.rmtree(repo, ignore_errors=True)
+  check_call(['git', 'clone', 'git@git:envoyproxy/go-control-plane', repo])
+  check_call(['git', '-C', repo, 'fetch'])
+  check_call(['git', '-C', repo, 'checkout', '-B', BRANCH, 'origin/master'])
+
+
+def findLastSyncSHA(repo):
+  # Determine last envoyproxy/envoy SHA in envoyproxy/go-control-plane
+  last_commit = check_output(
+      ['git', '-C', repo, 'log', '--grep=' + MIRROR_MSG, '-n', '1',
+       '--format=%B']).decode().strip()
+  # Initial SHA from which the APIs start syncing. Prior to that it was done manually.
+  if last_commit == "":
+    return 'e7f0b7176efdc65f96eb1697b829d1e6187f4502'
+  m = re.search(MIRROR_MSG + '(\w+)', last_commit)
+  return m.group(1)
+
+
+def updatedSinceSHA(repo, last_sha):
+  # Determine if there are changes to api since last SHA
+  changes = check_output(['git', 'rev-list', '%s..HEAD' % last_sha, 'api/envoy']).decode().split()
+  return changes
+
+
+def syncGoProtobufs(output, repo):
+  # Sync generated content against repo and return true if there is a commit necessary
+  dst = os.path.join(repo, 'envoy')
+  # Remove subtree at envoy in repo
+  shutil.rmtree(dst, ignore_errors=True)
```

Comment:
FWIW, I think these `rmtree` lines are quite dangerous, but since this is generally only going to run in CI, it's probably fine.

Reply:
Removed them. In CI, the tree is clean, so it's not necessary.